### ISSUE TITLE
Create osgi.service capability if use of BundleContext.registerService is detected

### ISSUE THREAD

**Issue by juergen-albert** [bndtools/bnd#4321]
If a class is annotated with `@Component` and has a `@Reference` we automatically create a Requirement of the declared service Interface. If the Component is a Service we also create a Capability by default. 

I would suggest to do the same if we detect any use of one of the known methods to register a Service manually. This is something most people forget or simply don't care/know about. We stumble upon this problem regularly in our own code and in the bundles of others on a regular basis. 

I'd like to provide an AnalyserPlugin that takes care of this issue. It should consider the nocapabilites and norequirements option similar to the DSAnnotations Analyser and would be active by default.

Does anybody sees any downsides to this? 

Do we have other methods beyond the register methods ones the `BundleContext` that should be considered?

**Comment by bjhargrave** [comment]
And how will Bnd know the name of the objectClass to use for the capability? It is  a lot of work and not always possible to grovel through the bytecode of a method body to find this.

**Comment by juergen-albert** [comment]
I know this might be a bit complicated, but we do something similar with Class.forName
https://github.com/bndtools/bnd/blob/8f33cf0047937e6d7875ddc3429aabf788500f26/biz.aQute.bndlib/src/aQute/bnd/osgi/Clazz.java#L871

I hoped to build on this and try to apply it to the BundleContext Methods.

**Comment by bjhargrave** [comment]
This seems only possible for the simple cases where a String or Class constant is passed to the registerService method. It is extremely difficult if the user passes a local variable or an expression or a String[].

**Comment by pkriens** [comment]
Why not create a warning when you see these register service calls and the class has no Service annotation? (Where Service annotation creates the capability.)

I am hesitant to automate too much here. These registration calls are very low level and should rarely be used.

**Comment by timothyjward** [comment]
> Why not create a warning when you see these register service calls and the class has no Service annotation? (Where Service annotation creates the capability.)

I agree with this - it will be much more reliable and the annotation is simple to add. Investing some time in an `@ServiceCapability` annotation might be worth it though, as the current syntax is a bit messy.

**Comment by bjhargrave** [comment]
> Investing some time in an `@ServiceCapability` annotation might be worth it though, as the current syntax is a bit messy.

Isn't this more of an OSGi annotation though?
